### ISSUE TITLE
[Proposal] API changes

### ISSUE THREAD

**Issue by kourge** [UrbanDoor/failable#12]
Problems:
- `isSuccess` is polymorphic, but having only one export of it forces us to overload it with exponential complexity. For example, its base type is `(x: any) => x is Success<any>`, but when used in conjunction with a `Failable<T>`, it is `<T>(x: Failable<T>) => x is Success<T>`. Note the addition of the type parameter! This doesn't scale if we continue to only export `isSuccess` once.
- `when` as a matcher is not one size fits all.
- the dispatching mechanism is not evenly applied to all union types.
- `when` doesn't always have a non-logging equivalent.

## Analysis
Making a library out of basic ADTs is hard, but so far the basic structure of `failable` is rather predictable:
- each core element (e.g. a success) consists of:
  - a type (`Success`)
  - a creator (`success`)
  - a predicate (`isSuccess`)
- each union type of multiple core elements comes with:
  - a matcher (e.g. `when`)
  - a logged matcher (e.g. `Failable.when`)
  - possibly a constructor (e.g. `Result.from`)

## Solution
The goal is to make all of these consistent and furthermore modular. The proposed changes are therefore:
- each union type must reside in its own submodule and must be "independently usable", i.e.
  - in `src`, a union type `Foo` must live in `src/foo`
  - from outside the library, it must be possible to `import ... from 'failable/foo'`
    - perhaps add `Failable` itself as an exemption so we don't have to write `import ... from 'failable/failable'`
- each union type must export all of its elements' types, creators, and predicate
  - for example, both `Failable` and `Result` use `Success<T>` and `Failure`
  - this must be done regardless of how those constituents are defined. as an implementation detail, perhaps `isFailure: (x: any) => x is Failure`, but if `Result` exports it, it must become `isFailure: (x: Result) => x is Failure`.
- each union type must export its own matcher
  - so the matcher exported by `Result` must expect a `Result` and not try to do anything fancy
  - sometimes coincidental compatibility can occur. For example, due to the way a structural type system works, `Failable`'s matcher can take a `Result`, but not vice versa
- a logged matcher and an raw matcher should be named differently, for example, call the raw one `match` and the logged one `when`. @squirly also expressed concern that `when` looks way too much like `then` and is likely a big source of confusion.
- the dispatch mechanic can be extracted into its own class, which can then be individually tested.

## Example
We export:
- `failable/core`
  - `type Success<T>`
    - `function success<T>(x: T): Success<T>`
    - `function isSuccess(x: any): x is Success<any>`
  - `type Failure`
    - `function failure(x: T): Failure`
    - `function isFailure(x: any): x is Failure`
- `failable/result`
  - `type Result<T>`
  - re-export `Success<T>` and pals, except:
    - `function isSuccess<T>(x: Result<T>): x is Success<T>`
  - re-export `Failure` and pals, except:
    - `function isFailure(x: Result<T>): x is Failure`
  - `function match<T, A, B>(x: Result<T>, options: {...}): A | B`, unlogged
  - `function when<T, A, B>(x: Result<T>, options: {...}): A | B`, logged
  - perhaps add `function isResult<T>(x: any): x is Result<T>`, which is a simple disjunction of all constituents (e.g. `isSuccess` and `isFailure`), which is also prone to false positives :frowning_face:
  - `namespace Result`
    - alias `match`
    - alias `when`
    - perhaps alias `is` to `isResult`
    - (insert dispatch crap here)
- repeat the same for `failable/failable`
- repeat the same for the upcoming `failable/loadable`

**Comment by squirly** [comment]
This makes sense to me.

The only change I could see making sense would be to have the `is*` methods being explicitly typed with each core type, `Pending | Success<T> | Failure | Retrying | Reloading<T>`. This would make these methods compatible with all super types. This would require making each of these types core types.

Either solution, the proposed or above, is fine with me.

**Comment by kourge** [comment]
Right, so look at how `isSuccess` is defined in `failable/core` using the broadest possible terms, but then it is re-exported out of `failable/result` with a much more specific type. This leads me to believe that all constituents should be defined in `failable/core` so that 1) they don't accidentally become compatible with other ones and 2) `failable/core` is the authoritative source of unrestrained type guards (guards that take `any`).
